### PR TITLE
ARROW-17640: [C++] Add File Handling Test cases for GlobFile handling in Substrait Read

### PR DESCRIPTION
**Pull request by vibhatha**
This PR adds test cases for GlobFile processing in `substrait::ReadRel`.

### CODE REVIEW

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 1]
```diff
@@ -3078,5 +3089,94 @@ TEST(Substrait, AggregateRelEmit) {
...
+              }, {
+                "i32": {}
+              }]
+            }
+          },
+          "local_files": {
+            "items": [
+              {
+                "uri_path_glob": ")" + glob_like_path +
+                                                                         R"(",
```
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/serde_test.cc:3166]: I formatted the code, but this is what it produces.

File: cpp/src/arrow/filesystem/filesystem_test.cc [hunk 2]
```diff
@@ -308,28 +308,27 @@ TEST(InternalUtil, GlobFiles) {
...
-  ASSERT_OK_AND_ASSIGN(infos, GlobFiles(fs, "A*/CD/?b*.txt"));
+  ASSERT_OK_AND_ASSIGN(infos, GlobFiles(fs, base_dir + "A*/CD/?b*.txt"));
   ASSERT_EQ(infos.size(), 2);
-  check_entries(infos, {"A/CD/ab.txt", "AB/CD/abc.txt"});
+  check_entries(infos, {base_dir + "A/CD/ab.txt", base_dir + "AB/CD/abc.txt"});
 
-  // Leading slash is optional but doesn't change behavior
-  ASSERT_OK_AND_ASSIGN(infos, GlobFiles(fs, "/A*/CD/?b*.txt"));
-  ASSERT_EQ(infos.size(), 2);
-  check_entries(infos, {"A/CD/ab.txt", "AB/CD/abc.txt"});
```
Comment (vibhatha) [inline cpp/src/arrow/filesystem/filesystem_test.cc:327]: cc @westonpace 
I removed this and updated the test case to function with and without leading `/`. Is this okay?

File: cpp/src/arrow/filesystem/path_util.h [hunk 3]
```diff
@@ -75,6 +75,9 @@ util::string_view RemoveTrailingSlash(util::string_view s);
 ARROW_EXPORT
 Status AssertNoTrailingSlash(util::string_view s);
 
+ARROW_EXPORT
+Status AssertLeadingSlash(util::string_view s);
```
Comment (vibhatha) [inline cpp/src/arrow/filesystem/path_util.h:79]: Should this be just return bool? It is a bit doubtful and tried to maintain the convention already in use.

cc @westonpace

**Comment by vibhatha** [comment]
cc @ArianaVillegas @sanjibansg

File: cpp/src/arrow/filesystem/path_util.h (hunk 3 above)
Comment (westonpace) [reply cpp/src/arrow/filesystem/path_util.h:79]: Yes, it should probably return a bool and be named `HasLeadingSlash` but, at that point, the method is just...

```
bool HasLeadingSlash(util::string_view key) {
  return key.front() == '/';
}
```

...and I'm not really sure such a method would be really needed.

**Review by westonpace — CHANGES_REQUESTED** [review]
Appreciate the extra test.  Just a few suggestions.

**Comment by vibhatha** [comment]
@westonpace Updated the PR and the `appveyor` is failing because of something related `Gandiva`.

**Review by westonpace — APPROVED** [review]
Thanks for the additional tests.

**Pull request merged**